libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteProgress.kt

```diff
@@ -0,0 +1,341 @@
+package com.mapbox.services.android.navigation.v5.routeprogress
+
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.api.directions.v5.models.LegStep
+import com.mapbox.api.directions.v5.models.RouteLeg
+import com.mapbox.geojson.Geometry
+import com.mapbox.geojson.Point
+import com.mapbox.navigator.BannerInstruction
+import com.mapbox.navigator.VoiceInstruction
+import com.mapbox.services.android.navigation.v5.utils.extensions.ifNonNull
+
+/**
+ * This class contains all progress information at any given time during a navigation session. This
+ * progress includes information for the current route, leg and step the user is traversing along.
+ * With every new valid location update, a new route progress will be generated using the latest
+ * information.
+ *
+ *
+ * The latest route progress object can be obtained through either the [ProgressChangeListener]
+ * or the [com.mapbox.services.android.navigation.v5.milestone.MilestoneEventListener] callbacks.
+ * Note that the route progress object's immutable.
+ *
+ *
+ * @since 0.1.0
+ */
+data class RouteProgress internal constructor(
+    var directionsRoute: DirectionsRoute? = null,
+    var legIndex: Int? = null,
+    var distanceRemaining: Double? = null,
+    var currentLegProgress: RouteLegProgress? = null,
+    var currentStepPoints: List<Point>? = null,
+    var upcomingStepPoints: List<Point>? = null,
+    var inTunnel: Boolean? = null,
+    var voiceInstruction: VoiceInstruction? = null,
+    var bannerInstruction: BannerInstruction? = null,
+    var currentState: RouteProgressState? = null,
+    var routeGeometryWithBuffer: Geometry? = null,
+    var currentStep: LegStep? = null,
+    var stepIndex: Int? = null,
+    var legDistanceRemaining: Double? = null,
+    var stepDistanceRemaining: Double? = null,
+    var legDurationRemaining: Double? = null,
+    var builder: Builder
+) {
+
+    /**
+     * Get the route the navigation session is currently using. When a reroute occurs and a new
+     * directions route gets obtained, with the next location update this directions route should
+     * reflect the new route. All direction route get passed in through
+     * [com.mapbox.services.android.navigation.v5.navigation.MapboxNavigation.startNavigation].
+     *
+     * @return a [DirectionsRoute] currently being used for the navigation session
+     * @since 0.1.0
+     */
+    fun directionsRoute() = directionsRoute
```

Comment:
already provided by `var directionsRoute`
here and below. don't need additional getters for properties. 
I think we should make fields in constructor `val` and if we'd like to change smt just call `copy()` like `object.copy(propertyA = "newValue")`

Reply:
I'm not sure I understand. This is a builder. and the method you are referencing is being used other places.

Reply:
this method is not a part of `Builder`, it's a part of data class